samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,370 @@
+        """
+        Creates boto client
+
+        Parameters
+        ----------
+        client_name: str
+            The name of the client
+        session: boto3.Session
+            The session created using customer config
+
+        Returns
+        -------
+        Service client instance
+        """
+        return get_boto_client_provider_from_session_with_config(session)(client_name)
+
+    def _compare_templates(self, template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        template_path : str
+            The template location of the current template
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        current_template = self.get_template(template_path)
+
+        if not current_template:
+            LOG.debug("Cannot obtain a working current template for template path %s", template_path)
+            return False
+
+        try:
+            last_deployed_template_str = self._cfn_client.get_template(
+                StackName=stack_name, TemplateStage="Original"
+            ).get("TemplateBody", "")
+        except ClientError as ex:
+            LOG.debug("Stack with name %s does not exist on CloudFormation", stack_name, exc_info=ex)
+            return False
+
+        last_deployed_template_dict = yaml_parse(last_deployed_template_str)
+
+        sanitized_current_template = copy.deepcopy(current_template)
+        sanitized_last_template = copy.deepcopy(last_deployed_template_dict)
+
+        sanitized_resources = self._remove_unnecessary_fields(sanitized_current_template)
+        self._remove_unnecessary_fields(sanitized_last_template, sanitized_resources)
+
+        if sanitized_last_template != sanitized_current_template:
+            LOG.debug("The current template is different from the last deployed version, we will not skip infra sync")
+            return False
+
+        # The recursive template check for Nested stacks
+        for resource_logical_id in current_template.get("Resources", {}):
+            resource_dict = current_template.get("Resources", {}).get(resource_logical_id, {})
+            resource_type = resource_dict.get("Type")
+
+            if resource_type in CODE_SYNCABLE_RESOURCES:
+                last_resource_dict = last_deployed_template_dict.get("Resources", {}).get(resource_logical_id, {})
+                if resource_type == AWS_LAMBDA_FUNCTION:
+                    if not resource_dict.get("Properties", {}).get("Code", None) == last_resource_dict.get(
+                        "Properties", {}
+                    ).get("Code", None):
+                        self._code_sync_resources.add(resource_logical_id)
```

Comment:
I think we should also check if they have local code path, otherwise we shouldn't trigger a code sync for them. Are we planning to check it in other place?

Reply:
I see, then we will need to use different templates, both package.template_file and package.output_template_file, one for confirming code sync, one for comparison

Reply:
Updated implementation to include one additional template file